woocommerce/woocommerce-ios
# Product List: product search UI

Fixes #1264 

- [ ] ⚠️ Update PR branch to `develop` before merging after base branch https://github.com/woocommerce/woocommerce-ios/pull/1289 is merged

Update release notes:


## Changes
- Created `ProductSearchViewController` similar to `OrderSearchViewController`
- Created `ProductSearchViewControllerStateCoordinator` to handle view controller state `ProductSearchViewControllerState`
  - Unit tests on state transitions
- Added search to the navigation bar left button in `ProductsViewController`, and navigated to `ProductSearchViewController` when tapped

## Testing
If possible, try testing with a store that has more than 25 products
- Launch the app
- Go to Products tab
- Tap Search icon on the left of the navigation bar --> observe the initial products, they should be the same as the products that have been loaded before on Products tab
- Enter a search query that has some results --> the products in the search results should contain the query
- Enter another query that has some results **that have not been fetched on Products tab yet** (this only works if the store has more than 25 products) --> the products containing the query but haven't been loaded should show up eventually
- Enter another query that has no results --> "No products found" placeholder should be shown
- Try switching to another store and repeat the steps above

## Example screenshots

no products | initial products before search | product search results
-- | -- | --
![Simulator Screen Shot - iPhone Xs - 2019-09-18 at 16 09 32](https://user-images.githubusercontent.com/1945542/65193707-e8c4ae80-da30-11e9-91ef-2482149ac014.png) | ![Simulator Screen Shot - iPhone Xs - 2019-09-19 at 10 00 25](https://user-images.githubusercontent.com/1945542/65264841-58d04480-dac4-11e9-9a5f-5866b30f65e2.png) | ![Simulator Screen Shot - iPhone Xs - 2019-09-19 at 09 59 59](https://user-images.githubusercontent.com/1945542/65264856-5ff75280-dac4-11e9-9671-aebfe141b3af.png)
